Allow to set bits in a range in bitseq

aboch opened the pull request:
Signed-off-by: Alessandro Boch <aboch@docker.com>

mrjana commented on bitseq/sequence.go:
```diff
@@ -186,12 +186,23 @@ func (h *Handle) getCopy() *Handle {
 
+// SetAnyInRange atomically sets the first unset bit in the specified range in the sequence and returns the corresponding ordinal
+func (h *Handle) SetAnyInRange(start, end uint32) (uint32, error) {
+	if end-start <= 0 || end > h.Bits() {
```
Shouldn't this be `end >= h.Bits()` since the bitseq is zero based?

aboch replied:
Good catch. Thanks.

mrjana commented on bitseq/sequence.go:
```diff
@@ -186,12 +186,23 @@ func (h *Handle) getCopy() *Handle {
 		return invalidPos, errNoBitAvailable
 	}
-	return h.set(0, true, false)
+	return h.set(0, 0, 0, true, false)
```
Can't you pass `h.Bits()` for `end` value so that you can avoid checking `end != 0` down below?

aboch replied:
Good suggestion. Let me give it a try.

mrjana commented on bitseq/sequence.go:
```diff
@@ -240,8 +251,11 @@ func (h *Handle) set(ordinal uint32, any bool, release bool) (uint32, error) {
-				bytePos, bitPos, err = getFirstAvailable(h.head)
+				bytePos, bitPos, err = getFirstAvailable(h.head, start)
 				ret = posToOrdinal(bytePos, bitPos)
+				if end != 0 && end < ret {
```
You don't need to check if `end != 0` if you always passed the right `end`

aboch replied:
I thought the inner function `GetFirstAvailable()` already checks for the upper limit `h.bits` and returned an error if we go over (this would happen if the number of bits is not a multiple of 32). But it doesn't...

I will make the change you suggested.
Thanks.

aboch commented:
@mrjana 
- Taken care of your comments. 
- Fixed a bug in `getFirstAvailable()` from the first diffs. 
- Updated few tests to test the `start`

PTAL when you get a chance.

mrjana commented:
LGTM

mavenugo commented:
@aboch for the changes in bitseq, without the tests that you have written, it is practically impossible to spot issues through code-review.
Having said that, the tests certainly helps and it LGTM :)

mavenugo commented on bitseq/sequence.go:
```diff
@@ -240,8 +251,11 @@ func (h *Handle) set(ordinal uint32, any bool, release bool) (uint32, error) {
-				bytePos, bitPos, err = getFirstAvailable(h.head)
+				bytePos, bitPos, err = getFirstAvailable(h.head, start)
 				ret = posToOrdinal(bytePos, bitPos)
+				if end < ret {
```
not a comment. but mostly a question. 
For the case of `(h *Handle) Set(ordinal uint32) error {`, the start and end are 0.
does `posToOrdinal` for this case return 0 ?
If not, this is obviously going to fail. can you please confirm ?

aboch replied:
It is not going to fail.
The range limits are evaluated/used only when the flag `any` is true. That's why I set them only during the `SetAny()` and `SetAnyInRange()` API calls. `h.set()` is the inner function which handles the four public API, it ends up having those many inputs. But being an internal function I did not see a problem or a refactor need, as long as it works.

aboch commented:
@mavenugo Understand, maybe in future I add more comments to the code and tests so that it may help understanding what the functions do.
